SplitFiles/customer.c

```diff
@@ -0,0 +1,168 @@
+
+#include "customer.h"
+#include "bread.h"
+#include "util.h"
+
+void redraw_customers(CustomerBase *base) {
+    unsigned char ypos = 4;
+    Customer* lastCustomer = base->rootCustomer;
+    while (lastCustomer->nextCustomer != NULL) {
+        lastCustomer = lastCustomer->nextCustomer;
+        printf(PRINTAT"\x12%c""%-12s", ypos, lastCustomer->breadOrder->desc);
+        ++ypos;
+    }
+    for (; ypos < 4 + MAX_CUSTOMER_COUNT; ++ypos) {
+        printf(PRINTAT"\x12%c""%-12s", ypos, "");
+    }
+}
+
+void customer_func(GameComponent* customers, GameParameters* params) {
+    CustomerBase* base = (CustomerBase*)(customers->ptr);
+    Customer* newCustomer = (Customer*)NULL;
+    int redraw = 0;
+
+    Customer* thisCustomer;
+    Customer* lastCustomer;
+
+    if (params->messageAddress == 20) {
+        //Someone sent toast!
+        BreadState* bread = (BreadState*)params->message;
+        thisCustomer = base->rootCustomer->nextCustomer;
+        lastCustomer = base->rootCustomer;
+        while (thisCustomer != NULL) {
+            if (bread->type->letter == thisCustomer->breadOrder->letter) {
+                //We have a match
+                params->score += bread->type->cost * 2 //SHOCKING markup
+                         + 100 - abs(bread->toastedness - 100) //100 toastedness gets 100 bonus points
+                         + thisCustomer->ticksLeft;  //Speedy toast = points!
+                (params->slices) += 1;
+                params->messageAddress = 0;
+                params->message = NULL;
+                params->messageSourceAddress = NULL;
+                //Give back memory
+                free(bread);
+                //Remove customer
+                base->customerCount -= 1;
+                lastCustomer->nextCustomer = thisCustomer->nextCustomer;
+                free(thisCustomer);
+                redraw = 1;
+                //end loop
+                thisCustomer = NULL;
+            } else {
+                //Step on
+                lastCustomer = thisCustomer;
+                thisCustomer = thisCustomer->nextCustomer;
+            }
+        }
+        if (!redraw) {
+            // No customer for this toast - bad
+            params->score -= (bread->type->cost + bread->toastedness);  //+ wasted energy cost
+            params->messageAddress = 0;
+            params->message = NULL;
+            params->messageSourceAddress = NULL;
+            free(bread);
+        }
+
+    }
+
+    thisCustomer = base->rootCustomer->nextCustomer;
+    lastCustomer = base->rootCustomer;
+    while (thisCustomer != NULL) {
+        thisCustomer->ticksLeft -= 1;
+        if (thisCustomer->ticksLeft <= 0) {
+            //Customer leaves in disgust - reset links
+            params->score -= thisCustomer->breadOrder->cost;
+            base->customerCount -= 1;
+            lastCustomer->nextCustomer = thisCustomer->nextCustomer;
+            free(thisCustomer);
+            thisCustomer = lastCustomer;
```

Comment:
redundant line given line 93?

Reply:
We just 'free'd thisCustomer - if we don't reassign it the freed memory will end up as lastCustomer at line 93

Reply:
OK, I *think* I follow this